src/backend/distributed/commands/non_main_db_distribute_object_ops.c

```diff
@@ -0,0 +1,450 @@
+ *   When creating a distributed object, we always have to mark such objects
+ *   as "distributed" but while for some object types we can delegate this to
+ *   main database, for some others we have to explicitly send a command to
+ *   all nodes in this code-path to achieve this. Callers need to implement
+ *   getMarkDistributedParams when that is the case.
+ *
+ *   Similarly when dropping a distributed object, we always have to unmark
+ *   the target object as "distributed" and our utility hook on remote nodes
+ *   achieve this via UnmarkNodeWideObjectsDistributed() because the commands
+ *   that we send to workers are executed via main db. However for the local
+ *   node, this is not the case as we're not in the main db. For this reason,
+ *   callers need to implement getUnmarkDistributedParams to unmark an object
+ *   for local node as well.
+ *
+ *   We don't expect both of these to be NULL at the same time. However, it's
+ *   okay if both of these are NULL.
+ *
+ *   Finally, while getMarkDistributedParams is expected to return "a list of
+ *   objects", this is not the case for getUnmarkDistributedParams. This is
+ *   because, while we expect that a drop command might drop multiple objects
+ *   at once, we don't expect a create command to create multiple objects at
+ *   once.
+ *
+ *  cannotBeExecutedInTransaction:
+ *   Indicates whether the statement cannot be executed in a transaction. If
+ *   this is set to true, the statement will be executed directly on the main
+ *   database because there are no transactional visibility issues for such
+ *   commands.
+ *
+ *   And when this is true, we don't expect getMarkDistributedParams and
+ *   getUnmarkDistributedParams to be implemented.
+ */
+typedef struct NonMainDbDistributeObjectOps
+{
+	MarkDistributedGloballyParams * (*getMarkDistributedParams)(Node * parsetree);
+	List * (*getUnmarkDistributedParams)(Node * parsetree);
+	bool cannotBeExecutedInTransaction;
+} NonMainDbDistributeObjectOps;
+
+
+/*
+ * getMarkDistributedParams and getUnmarkDistributedParams callbacks for
+ * NonMainDbDistributeObjectOps.
+ */
+static MarkDistributedGloballyParams * CreateRoleStmtGetMarkDistributedParams(
+	Node *parsetree);
+static List * DropRoleStmtGetUnmarkDistributedParams(Node *parsetree);
+
+
+/*
+ * NonMainDbDistributeObjectOps for different command types.
+ *
+ * Naming of these structs are stolen from distribute_object_ops.c. We use
+ * "Any" if the command doesn't have any other variations or if the struct
+ * implements it for all its variations. For example, while we name the struct
+ * for CreateRoleStmt as Any_CreateRole, we name the struct that implements
+ * SecLabelStmt for role objects as Role_SecLabel.
+ */
+static const NonMainDbDistributeObjectOps *operationArray[] = {
+	[T_CreateRoleStmt] = &(NonMainDbDistributeObjectOps) {
+		.getMarkDistributedParams = CreateRoleStmtGetMarkDistributedParams,
+		.getUnmarkDistributedParams = NULL,
```

Comment:
I'm not a huge fan of the `getMarkDistributedParams` or `getUnmarkDistributedParams` abstractions. Since both functions are only needed for one. It seems easier to just check the NodeType in the calling function and call the relevant function for these two cases. Which means we don't get this extra layer of indirection.

my rule of thumb: If an abstraction is only used once, then it's probably the wrong abstraction.

Reply:
I agree with you in that if an abstraction is used only once, it's probably not a good idea. However, I also don't much want to have a command-specific logic to mark/unmark objects distributed in those pre/postprocess functions.

So if you don't have a strong objection, I'd like to move forward with the commit you pushed and would like to not remove those two callbacks.

Reply:
I think moving command specific logic out of the post/preprocess functios is a good idea. I just don't think it should be done in a generic way. Creating generic code for stuff that you do once, pretty much always means that in the future you have to change the way you do it anyway. Because you chose the wrong way of making it generic.
```c
if (nodeType(stmt) == T_CreateUser) {
    CreateRoleStmtMarkDistributed()
}
```

So to be clear: yes, to functions moving stuff out of call site. no, to adding single-use callbacks to `NonMainDbDistributeObjectOps`.

Reply:
For reference this blogpost describes my worry about having the wrong abstraction much better than I could myself: https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction

Reply:
done